clustering.Rmd

```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
-before. This is because we need to iterate over `k`, which is the *second argument* 
-to the `kmeans` function. In the past, we have used `map` only to iterate over values of the
-*first argument* of a function. Since that is the default, we could simply 
-write `map(data_frame, function_name)`. This won't work here; we need to 
-provide our data frame as the first argument to the `kmeans` function.
-
-The solution is to create something called an *anonymous function*. \index{function!anonymous}
-An anonymous function is a function that has no name,
-unlike other functions you've seen so far (`kmeans`, `select`, etc). 
-To do this we will write our map statement like this:
-
-```{r, eval = FALSE}
-map(penguin_clust_ks, function(k) kmeans(scaled_data, k))
+Then we use `rowwise` \index{rowwise} + `mutate` to apply the `kmeans` function 
+within each row, to each K. 
+However, given that the `kmeans` function 
+returns a model object to us, (not a vector),
+we will need to store the results as a list columm.
+This can work because lists are super vectors, and so they are legitimate 
+data structures for data frame columns. 
+To make this work, 
+we have to put each model object in a list using the `list` function.
+We demonstrate how to do this below:
+
+```{r}
+penguin_clust_ks <- tibble(k = 1:9) |>
+  rowwise() |>
+  mutate(penguin_clusts = list(kmeans(standardized_data, k)))
 ```
 
-The anonymous function in the above call is `function(k) kmeans(scaled_data, k)`.
-This function takes a single argument (`k`) and evaluates `kmeans(scaled_data, k)`.
-Since `k` is the *first* (and only) argument to the function, we can use `map` just
-like we did before! The rest of the call above does just that -- it passes each row of
-`penguin_clust_ks` to our anonymous function.
+If we take a look at our data frame `penguin_clust_ks` now, 
+we see that it has two columns: one with the value for K, 
+and the other holding the clustering model object in a list column.
 
-Below, we execute this `map` call inside of a `mutate` call on the 
-`penguin_clust_ks` data frame and get a list column that contains a K-means
-clustering object for each value of K we had:
-\index{mutate}
-
-```{r 10-choose-k-part2}
-penguin_clust_ks <- tibble(k = 1:9) |>
-  mutate(penguin_clusts = map(k, function(ks) kmeans(scaled_data, ks)))
+```{r}
 penguin_clust_ks
 ```
 
-Next, we use `map` again to apply `glance` \index{glance} to each of the K-means
-clustering objects to get the clustering statistics (including WSSD). The output
-of glance is a data frame, and so we get another list column. This results in a 
-complex data frame with 3 columns, one for K, one for the 
+> If we wanted to get one of the clusterings out 
+> of the list column in the data frame,
+> we can use our familiar friends `slice` and `pull`.
+> However, before we do this, we need to turn `rowwise`.
+> To do that, we use `ungroup` 
+> (predictably, this can also turn `group_by` off as well).
```

Comment:
ah, I guess it should be "turn rowwise off" 

I feel like there must be a better way to say these things than "turn off". Disable? Deactivate?

Reply:
Actually, we don't need to turn `rowwise` off here (another bug caused me to think this was that, but after fixing that, I just tested it and we don't need to do it. So removing it here.